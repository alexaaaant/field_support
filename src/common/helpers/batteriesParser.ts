import { Academies, BatteriesRaw, Devices } from '../../types/batteries';

export const batteriesParser = (data: BatteriesRaw) => {
    const devices: Devices = {};
    const academies: Academies = {};
    const academiesPrioritized: number[] = [];
  
    data.forEach(entry => {
      if (!devices[entry.serialNumber]) {
        devices[entry.serialNumber] = {
          academyId: entry.academyId,
          serialNumber: entry.serialNumber,
          measurements: [],
          wearLevel: 0,
        };
      }
      devices[entry.serialNumber].measurements.push({
        batteryLevel: entry.batteryLevel,
        timestamp: +new Date(entry.timestamp),
      });
    });
  
    Object.values(devices).forEach(device => {
      device.measurements.sort((a, b) => a.timestamp - b.timestamp);
      let totalBatteryUsage = 0;
      let count = 0;
  
      for (let i = 1; i < device.measurements.length; i++) {
        const previous = device.measurements[i - 1];
        const current = device.measurements[i];
        const timeDifference = (current.timestamp - previous.timestamp) / (1000 * 60 * 60 * 24);
        const batteryUsage = previous.batteryLevel - current.batteryLevel;
  
        if (batteryUsage > 0) {
          totalBatteryUsage += batteryUsage / timeDifference;
          count++;
        }
      }
      const wearLevel = totalBatteryUsage / count;
      device.wearLevel = wearLevel;

      if (count > 0 && wearLevel > 0.3) {
        if (!academies[device.academyId]) {
          academies[device.academyId] = { batteryIssues: 0, devices: [] };
          academiesPrioritized.push(device.academyId);
        }
        academies[device.academyId].batteryIssues++;
        academies[device.academyId].devices.push(device);
      }
    });
    academiesPrioritized.sort((a, b) => academies[b].batteryIssues - academies[a].batteryIssues);
    return { academies, academiesPrioritized };
  
};