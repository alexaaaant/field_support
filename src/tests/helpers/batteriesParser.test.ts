import { batteriesParser } from '../../common/helpers/batteriesParser';
import { badBatteries, goodBatteries } from './mockData';

describe('batteriesParser', () => {
    it('should return an empry array, if all batteries are good', () => {
        const result = batteriesParser(goodBatteries);

        expect(result).toStrictEqual({'academies': {}, 'academiesPrioritized': []});
    });

    it('should return bad batteries, if some batteries are bad', () => {
        const result = batteriesParser(badBatteries);

        expect(result).toMatchSnapshot();
    });
});
