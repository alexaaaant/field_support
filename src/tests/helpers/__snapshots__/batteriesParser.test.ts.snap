// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`batteriesParser should return bad batteries 1`] = `
Object {
  "academies": Object {
    "30007": Object {
      "batteryIssues": 1,
      "devices": Array [
        Object {
          "academyId": 30007,
          "measurements": Array [
            Object {
              "batteryLevel": 0.98,
              "timestamp": 1558075835158,
            },
            Object {
              "batteryLevel": 0.6,
              "timestamp": 1558076228290,
            },
          ],
          "serialNumber": "1805C67HD02009",
          "wearLevel": 83.51393425109123,
        },
      ],
    },
    "30008": Object {
      "batteryIssues": 1,
      "devices": Array [
        Object {
          "academyId": 30008,
          "measurements": Array [
            Object {
              "batteryLevel": 0.98,
              "timestamp": 1558075835158,
            },
            Object {
              "batteryLevel": 0.7,
              "timestamp": 1558076228290,
            },
          ],
          "serialNumber": "1805C67HD02019",
          "wearLevel": 61.53658313238302,
        },
      ],
    },
  },
  "academiesPrioritized": Array [
    30007,
    30008,
  ],
}
`;

exports[`batteriesParser should return bad batteries, if some batteries are bad 1`] = `
Object {
  "academies": Object {
    "30007": Object {
      "batteryIssues": 1,
      "devices": Array [
        Object {
          "academyId": 30007,
          "measurements": Array [
            Object {
              "batteryLevel": 0.98,
              "timestamp": 1558075835158,
            },
            Object {
              "batteryLevel": 0.6,
              "timestamp": 1558076228290,
            },
          ],
          "serialNumber": "1805C67HD02009",
          "wearLevel": 83.51393425109123,
        },
      ],
    },
    "30008": Object {
      "batteryIssues": 1,
      "devices": Array [
        Object {
          "academyId": 30008,
          "measurements": Array [
            Object {
              "batteryLevel": 0.98,
              "timestamp": 1558075835158,
            },
            Object {
              "batteryLevel": 0.7,
              "timestamp": 1558076228290,
            },
          ],
          "serialNumber": "1805C67HD02019",
          "wearLevel": 61.53658313238302,
        },
      ],
    },
  },
  "academiesPrioritized": Array [
    30007,
    30008,
  ],
}
`;
